picard/ui/searchdialog.py

```diff
@@ -0,0 +1,466 @@
+        self.setWindowTitle(_("Track Search Results"))
+        self.table_headers = [
+                _("Name"),
+                _("Length"),
+                _("Artist"),
+                _("Release"),
+                _("Date"),
+                _("Country"),
+                _("Type")
+        ]
+
+    def search(self, text):
+        """Performs search using query provided by the user."""
+        self.retry_params = (self.search, text)
+        self.search_box.search_edit.setText(text)
+        self.show_progress()
+        self.tagger.xmlws.find_tracks(self.handle_reply,
+                query=text,
+                search=True,
+                limit=25)
+
+    def load_similar_tracks(self, file_):
+        """Performs search by using existing metadata information
+        from the file."""
+        self.retry_params = (self.load_similar_tracks, file_)
+        self.file_ = file_
+        metadata = file_.orig_metadata
+        query = {
+                'track': metadata['title'],
+                'artist': metadata['artist'],
+                'release': metadata['album'],
+                'tnum': metadata['tracknumber'],
+                'tracks': metadata['totaltracks'],
+                'qdur': str(metadata.length / 2000),
+                'isrc': metadata['isrc'],
+        }
+        if config.setting["use_adv_search_syntax"]:
+            # Display the query in advance syntax format.
+            query_str = ' '.join(['%s:(%s)' % (item, value) for item, value in query.iteritems() if value])
+        else:
+            # Display only the track title
+            query_str = query["track"]
+        # `query_str` is used only for presenting purpose. Actual query consists of all filters and follows
+        # advanced query syntax.
+        query["limit"] = 25
+        self.search_box.search_edit.setText(query_str)
+        self.show_progress()
+        self.tagger.xmlws.find_tracks(
+                self.handle_reply,
+                **query)
+
+    def retry(self):
+        """Retries the search using information from `retry_params`.
+
+        `retry_params` is a tuple having search information.
+        retry_params[0] -- Method to be used to perform search
+                        -- Can be `self.search()` or `self.load_similar_tracks()`
+        retry_params[1] -- Search query information
+                        -- Can be a text string, or a File object
+        """
+        self.retry_params[0](self.retry_params[1])
+
+    def handle_reply(self, document, http, error):
+        if error:
+            self.network_error(http, error)
+            return
+
+        try:
+            tracks = document.metadata[0].recording_list[0].recording
+        except (AttributeError, IndexError):
+            self.no_results_found()
+            return
+
+        if self.file_:
+            # Sort the results by comparing them to original metadata tags
+            tmp = sorted((self.file_.orig_metadata.compare_to_track(
```

Comment:
If you rename `tmp` to `sorted_results`, everyone knows what it is and you can get rid of the comment above.

I'd also format this a bit differently to make it clear which argument belongs to which function (although the first line now looks ugly):

```python
            sorted_results = sorted(
                (self.file_.orig_metadata.compare_to_track(
                    track,
                    File.comparison_weights)
                 for track in tracks),
                reverse=True,
                key=itemgetter(0))

```